server/src/main/java/org/opensearch/action/support/replication/ReplicationProxyRequest.java

```diff
@@ -0,0 +1,135 @@
+        ShardRouting primaryRouting,
+        long globalCheckpoint,
+        long maxSeqNoOfUpdatesOrDeletes,
+        PendingReplicationActions pendingReplicationActions,
+        ReplicaRequest replicaRequest
+    ) {
+        this.replicationModeAwareShardRouting = Objects.requireNonNull(replicationModeAwareShardRouting);
+        this.primaryRouting = Objects.requireNonNull(primaryRouting);
+        this.globalCheckpoint = globalCheckpoint;
+        this.maxSeqNoOfUpdatesOrDeletes = maxSeqNoOfUpdatesOrDeletes;
+        this.pendingReplicationActions = Objects.requireNonNull(pendingReplicationActions);
+        this.replicaRequest = Objects.requireNonNull(replicaRequest);
+    }
+
+    public ReplicationModeAwareShardRouting getReplicationModeAwareShardRouting() {
+        return replicationModeAwareShardRouting;
+    }
+
+    public ShardRouting getPrimaryRouting() {
+        return primaryRouting;
+    }
+
+    public long getGlobalCheckpoint() {
+        return globalCheckpoint;
+    }
+
+    public long getMaxSeqNoOfUpdatesOrDeletes() {
+        return maxSeqNoOfUpdatesOrDeletes;
+    }
+
+    public PendingReplicationActions getPendingReplicationActions() {
+        return pendingReplicationActions;
+    }
+
+    public ReplicaRequest getReplicaRequest() {
+        return replicaRequest;
+    }
+
+    /**
+     * Builder of ReplicationProxyRequest.
+     *
+     * @opensearch.internal
+     */
+    public static class Builder<ReplicaRequest> {
+
+        private ReplicationModeAwareShardRouting replicationModeAwareShardRouting;
+        private ShardRouting primaryRouting;
+        private long globalCheckpoint;
+        private long maxSeqNoOfUpdatesOrDeletes;
+        private PendingReplicationActions pendingReplicationActions;
+        private ReplicaRequest replicaRequest;
+
+        public Builder() {}
+
+        public Builder<ReplicaRequest> setReplicationModeAwareShardRouting(
+            ReplicationModeAwareShardRouting replicationModeAwareShardRouting
+        ) {
+            this.replicationModeAwareShardRouting = replicationModeAwareShardRouting;
+            return this;
+        }
+
+        public Builder<ReplicaRequest> setPrimaryRouting(ShardRouting primaryRouting) {
+            this.primaryRouting = primaryRouting;
+            return this;
+        }
+
+        public Builder<ReplicaRequest> setGlobalCheckpoint(long globalCheckpoint) {
+            this.globalCheckpoint = globalCheckpoint;
+            return this;
+        }
+
+        public Builder<ReplicaRequest> setMaxSeqNoOfUpdatesOrDeletes(long maxSeqNoOfUpdatesOrDeletes) {
+            this.maxSeqNoOfUpdatesOrDeletes = maxSeqNoOfUpdatesOrDeletes;
+            return this;
+        }
+
+        public Builder<ReplicaRequest> setPendingReplicationActions(PendingReplicationActions pendingReplicationActions) {
+            this.pendingReplicationActions = pendingReplicationActions;
+            return this;
+        }
```

Comment:
Consider moving this to the constructor of the `Builder` since all these are mandatory.